doc/designs/fips_indicator.md

```diff
@@ -0,0 +1,262 @@
+- Cons: The query can only be done after the output is set.
+- Cons: The indicator code would end up having to be set in different places depending on the algorithm after the output is finalized. This would be fairly messy as the point where it is called is set could be different for different algorithms.
+- Cons: The locking increment seems messy.
+
+### Proposed Solution (Using an indicator everywhere)
+
+Add a OSSL_PARAM getter to each provider algorithm context.
+By default if the getter is not handled then it would return not approved.
+
+- Pros: The code is easier to find since it is part of the get_ctx_params function.
+- Pros: The getter can be called at any point after the setting is done.
+
+Any fips algorithm that is approved would then need a setter that at a minimum contains code similar to the following
+
+``` C
+int ossl_xxx_fips_approved(void)
+{
+ifdef FIPS_MODULE
+    return 1; // conditional code would go here for each algorithm if required *
+else
+    return 0;
+endif
+}
+```
+
+and in the algorithms get_ctx() function
+
+``` C
+int xxx_get_fips_approved(OSSL_PARAM params[])
+{
+    p = OSSL_PARAM_locate(params, OSSL_FIPS_PARAM_APPROVED);
+    if (p != NULL && !OSSL_PARAM_set_int(p, ossl_xxx_fips_approved()))
+        return 0;
+    return 1;
+}
+```
+
+### API’s that would be used to support this are
+
+- EVP_PKEY Keygen, Encryption, Signatures, Key Exchange, KEM
+
+``` C
+EVP_PKEY_CTX_get_params(ctx, );
+```
+
+(Note that this would mean you could not use functions that hide the ctx such as EVP_PKEY_Q_keygen()!)
+
+- Ciphers
+
+``` C
+EVP_CIPHER_CTX_get_params()
+```
+
+- Digests
+
+``` C
+EVP_MD_CTX_get_params()
+```
+
+- KDF’s
+
+``` C
+EVP_KDF_CTX_get_params()
+```
+
+- MAC’s
+
+``` C
+EVP_MAC_CTX_get_params()
+```
+
+- RAND
+
+``` C
+EVP_RAND_CTX_get_params()
+```
+
+### Backwards Compatibility
+
+Previous providers do not support this operation, so they will return not approved if they are not handled.
+
+### Alternate Solution
+
+If we had different kinds of compliance requirements (something other than FIPS) either a separate getter could be added or the getter could return a int type instead of a 0 or 1..
+(e.g 1 = fips approved, 2 = some other compliance approved)
+
+Changes Required for indicators
+-------------------------------
+
+### key size >= 112 bits
+
+There are a few places where we do not enforce key size that need to be addressed.
+
+- HMAC  Which applies to all algorithms that use HMAC also.
+- CMAC
+
+### Algorithm Transitions
+
+Should we remove these algorithms completely from the fips provider, or use indicators?
+
+- DES_EDE3_ECB.  Disallowed for encryption, allowed for legacy decryption
+- DSA.  Keygen and Signing are no longer approved, not sure if verify is still approved.
+- RSA Signing using PKCSV15.  Rsa self test for sign may need to change as the default pad_mode is PKCSV15. Check if saltlen needs a indicator. Padding mode updates required in rsa_check_padding(). Check if sha1 is allowed?
```

Comment:
Could you clarify where this requirement comes from? I am only aware of PKCS#1 v1.5 padding being removed for RSA encryption/decryption.

Reply:
Yes, SP 800-131Ar2 has a stricter limit for RSA KEM and OAEP because SP 800-56Br2 starts at 2048.

Reply:
Which means the check I currently do is wrong..